File: camera/src/provisioning.rs
use std::path::Path;
use std::time::Duration;

use anyhow::Result;
use tokio_util::sync::CancellationToken;

use crate::http_client;
use crate::qr_enrollment::QrResult;

/// Run the provisioning flow: scan QR → configure WiFi → provision with server.
///
/// Blocks until provisioning succeeds or is cancelled.
/// Returns (api_key, device_id, server_url).
pub async fn run_provisioning(
    data_dir: &Path,
    device_serial: &str,
    cancel: CancellationToken,
) -> Result<(String, String, String)> {
    tracing::info!("entering provisioning mode — waiting for QR code");

    // Check for pre-provisioned token file (Docker auto-provision)
    let token_path = data_dir.join("provision_token");
    let server_url_path = data_dir.join("server_url");
    if token_path.exists() && server_url_path.exists() {
        let token = tokio::fs::read_to_string(&token_path).await?;
        let server_url = tokio::fs::read_to_string(&server_url_path).await?;
        let token = token.trim();
        let server_url = server_url.trim();

        if !token.is_empty() && !server_url.is_empty() {
            tracing::info!("found pre-provisioned token file, attempting provisioning");
            match http_client::provision(server_url, token, device_serial).await {
                Ok(resp) => {
                    http_client::save_credentials(
                        data_dir,
                        &resp.api_key,
                        &resp.device_id,
                        server_url,
                    )?;
                    let _ = tokio::fs::remove_file(&token_path).await;
                    return Ok((resp.api_key, resp.device_id, server_url.to_string()));
                }
                Err(e) => {
                    tracing::warn!("pre-provisioned token failed: {e}, falling through to QR");
                }
            }
        }
    }

    // QR scanning loop
    loop {
        if cancel.is_cancelled() {
            anyhow::bail!("provisioning cancelled");
        }

        match scan_and_provision(data_dir, device_serial).await {
            Ok(creds) => return Ok(creds),
            Err(e) => {
                tracing::warn!("provisioning attempt failed: {e}");
                tokio::select! {
                    _ = tokio::time::sleep(Duration::from_secs(5)) => {}
                    _ = cancel.cancelled() => anyhow::bail!("provisioning cancelled"),
                }
            }
        }
    }
}

/// Single attempt: scan a QR code, extract server URL + token, provision.
async fn scan_and_provision(
    data_dir: &Path,
    device_serial: &str,
) -> Result<(String, String, String)> {
    let qr_result = crate::qr_enrollment::scan_for_claim_token().await?;

    let (token, server_url) = match qr_result {
        QrResult::Wifi {
            ssid,
            claim_token,
            server_addr,
        } => {
            // WiFi QR — configure WiFi, extract token
            tracing::info!(ssid = %ssid, "configuring WiFi from QR code");
            if let Err(e) = crate::network::ensure_wifi(&ssid, None).await {
                tracing::warn!("WiFi configuration failed: {e}");
            }
            tokio::time::sleep(Duration::from_secs(3)).await;

            let token = claim_token.ok_or_else(|| {
                anyhow::anyhow!("QR code has WiFi but no provisioning token")
            })?;
            let server_url = server_addr.ok_or_else(|| {
                anyhow::anyhow!("QR code has no server URL")
            })?;
            (token, server_url)
        }
        QrResult::ClaimToken { token, server_addr } => {
            let server_url = server_addr.ok_or_else(|| {
                anyhow::anyhow!("QR code has no server URL")
            })?;
            (token, server_url)
        }
    };

    // The server_addr from the old QR format is host:port (QUIC).
    // For the new flow, the `s` field should be a full HTTPS URL.
    // Support both: if it looks like a URL, use as-is; otherwise, wrap it.
    let server_url = if server_url.starts_with("http://") || server_url.starts_with("https://") {
        server_url
    } else {
        format!("https://{server_url}")
    };

    tracing::info!(server = %server_url, "provisioning with server");
    let resp = http_client::provision(&server_url, &token, device_serial).await?;

    http_client::save_credentials(data_dir, &resp.api_key, &resp.device_id, &server_url)?;

    tracing::info!(device_id = %resp.device_id, "provisioning complete");

    Ok((resp.api_key, resp.device_id, server_url))
}

File: camera/provisioning.go
package main

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"

	"github.com/cargocam/ghostcam/common"
)

// RunProvisioning provisions the camera via a one-time token. Token and
// server URL are resolved in order:
//
//   1. CLI/env (cfg.ProvisionToken, cfg.ServerURL)
//   2. Flat files ({dataDir}/provision_token, {dataDir}/server_url)
//   3. QR scan via rpicam-still AND BLE GATT peripheral, raced — first
//      payload wins, loser's context is cancelled. 5-min shared timeout.
//
// The camera sends its ed25519 public key to the server — no secret is
// returned. Returns nil credentials if no token is available from any source.
func RunProvisioning(ctx context.Context, cfg *CameraConfig, deviceSerial string, identity *Identity) (*Credentials, error) {
	token, serverURL := resolveProvisionInputs(cfg)

	// Fallback: race QR scan + BLE GATT peripheral.
	if token == "" || serverURL == "" {
		payload, err := raceQRandBT(ctx, identity.DeviceID)
		if err != nil {
			return nil, err
		}
		if payload == nil {
			slog.Info("no provision token available, waiting for provisioning")
			return nil, nil
		}
		token = payload.Token
		serverURL = payload.Server

		if payload.WifiSSID != "" {
			psk := payload.WifiPassword
			if err := EnsureWifi(ctx, payload.WifiSSID, &psk); err != nil {
				slog.Warn("WiFi from onboarding payload failed", "ssid", payload.WifiSSID, "err", err)
			}
			WaitForRoute(ctx)
		}
	}

	// Ensure server_url is a full URL
	if !strings.HasPrefix(serverURL, "http://") && !strings.HasPrefix(serverURL, "https://") {
		serverURL = "https://" + serverURL
	}

	slog.Info("attempting provisioning", "server", serverURL)

	if err := Provision(ctx, serverURL, token, deviceSerial, identity); err != nil {
		return nil, err
	}

	creds := &Credentials{
		DeviceID:  identity.DeviceID,
		ServerURL: serverURL,
		Identity:  identity,
	}

	if err := SaveCredentials(cfg.DataDir, creds); err != nil {
		return nil, err
	}

	// Remove the one-time token file (if it exists)
	_ = os.Remove(filepath.Join(cfg.DataDir, "provision_token"))

	slog.Info("provisioning complete", "device_id", creds.DeviceID)
	return creds, nil
}

// raceQRandBT runs ScanQR and ScanBT concurrently and returns the first
// payload that arrives, cancelling the loser. nil result + nil error
// means both timed out without a payload.
func raceQRandBT(ctx context.Context, deviceID string) (*common.QRPayload, error) {
	raceCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	type result struct {
		payload *common.QRPayload
		err     error
		source  string
	}
	ch := make(chan result, 2)

	go func() {
		p, err := ScanQR(raceCtx)
		ch <- result{p, err, "qr"}
	}()
	go func() {
		// Device-ID prefix used in the BT advertised name so users see
		// e.g. "Ghostcam-45a8b310" and can disambiguate when multiple
		// un-onboarded Pis are nearby.
		prefix := deviceID
		if len(prefix) > 8 {
			prefix = prefix[:8]
		}
		p, err := ScanBT(raceCtx, prefix)
		ch <- result{p, err, "bt"}
	}()

	// Wait for either a payload or both to finish empty.
	var lastErr error
	for i := 0; i < 2; i++ {
		r := <-ch
		if r.payload != nil {
			slog.Info("onboarding payload accepted", "source", r.source)
			cancel()
			// Drain the second goroutine in the background so it can exit
			// cleanly on its own; don't block the caller.
			go func() { <-ch }()
			return r.payload, nil
		}
		if r.err != nil && !errors.Is(r.err, context.Canceled) {
			lastErr = fmt.Errorf("%s: %w", r.source, r.err)
		}
	}
	return nil, lastErr
}

// resolveProvisionInputs returns (token, serverURL) by merging CLI/env
// and flat file sources. Either source can provide either value.
func resolveProvisionInputs(cfg *CameraConfig) (string, string) {
	token := cfg.ProvisionToken
	serverURL := cfg.ServerURL

	if token == "" {
		token = readTrimmedFile(filepath.Join(cfg.DataDir, "provision_token"))
	}
	if serverURL == "" {
		serverURL = readTrimmedFile(filepath.Join(cfg.DataDir, "server_url"))
	}

	if token != "" && serverURL != "" {
		return token, serverURL
	}
	return "", ""
}
